### DISCUSSION TITLE
Rendering pyvista in myst markdown websites

### DISCUSSION THREAD

**Discussion by stevejpurves** [pyvista/pyvista#6189] [Q&A] [answered]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

MyST Markdown let's people build interactive websites from a combination of jupyter notebooks and markdown files. 

We are running into problems mainly as the website frontend is on a different domain to the jupyter server that pyvista is interacting with.

Or we need to take a different approach altogether, we're looking for guidance on our current approach and if there is a better/alternative way to approach this, we'd love to know.

see example here: https://cn-testing-staging.curve.space/articles/018fde2b-07b8-7ae6-9d0c-fa22325c6a04/d-plots#main (press power button on figure)

### Description

MyST Markdown let's people build interactive websites from a combination of jupyter notebooks and markdown files. These websites can be hosted and pointed to BinderHub instances that can launch a jupyter server to allow notebooks (or interactive figures made by embedding a notebook output) on the website to be executed.

(Note: Myst Markdown renders jupyterlab (4.0) output areas and has ipywidgets support.)

We are trying to get pyvista to work in myst websites.

We have a custom BinderHub deployment and have been working though various issues to try and get this to work.

Firstly, we encountered a problem on relative paths. When pyvista loads it attempts to load an `index.html` file in an iframe, from the jupyter user server on a relative url `/user/aaaaa-bbbb-cccc/...` which 404s as it's on another domain. I can see how this would work fine in juptyerlab.

We go around this by exposing an environment variable with the jupyterhub domain and setting:

```
cn_prefix = os.environ.get("CN_JUPYTER_SERVER_BASE_URL",None)
jh_prefix = os.environ.get("JUPYTERHUB_SERVICE_PREFIX",None)
if cn_prefix is not None and jh_prefix is not None:
    pv.global_theme.trame.server_proxy_prefix = cn_prefix + jh_prefix + "proxy/"
```

We then hit a cors issue with frame ancestors. We resolved this by updating the jupyterhub cors policy.

We then hit the last issue, where the call to the jupyter server does not have the correct token (maybe makes sense as it's in a cookie, and the iframe can't access that (different domain)? see 👇

![image](https://github.com/pyvista/pyvista/assets/1473646/7179e5e4-caae-408f-8f96-a5423c834f6f)

We expect to see the pyvista plot in the LHS panel. I don't think there is a way for us to get the token on to the iframe url, so this may be the end of the line for this approach.

Overall my questions are:

* Is there a way to get tht token information, or otherwise another way influence the url being constructed for that iframe?
* Is there a totally different/better approach using pyvista that we could look at here? we have full control over both the website hosting and BinderHub but they are on separate domains?

Thanks in advance for any insights / help!

### System Information

```shell
Architecture : 64bit
                 RAM : 31.0 GiB
         Environment : Jupyter
          GPU Vendor : Mesa/X.org
        GPU Renderer : llvmpipe (LLVM 15.0.6, 256 bits)
         GPU Version : 4.5 (Core Profile) Mesa 22.3.2
    MathText Support : True

  Python 3.10.11 | packaged by conda-forge | (main, May 10 2023, 18:58:44)
  [GCC 11.3.0]

             pyvista : 0.43.8
                 vtk : 9.3.0
               numpy : 1.26.4
          matplotlib : 3.8.4
              scooby : 0.10.0
               pooch : 1.8.1
              pillow : 10.3.0
             imageio : 2.34.1
             IPython : 8.13.2
            colorcet : 3.1.0
             cmocean : 4.0.3
          ipywidgets : 8.1.2
               scipy : 1.13.1
                tqdm : 4.65.0
              meshio : 5.3.5
          jupyterlab : 4.2.1
               trame : 3.6.0
        trame_client : 3.0.2
        trame_server : 3.0.0
           trame_vtk : 2.8.8
       trame_vuetify : 2.5.0
jupyter_server_proxy : 4.1.2
        nest_asyncio : 1.5.6
```

**Comment by tkoyama010** [comment]
Thank you for using PyVista in such a awesome project. I also love Myst and JupyterBook etc. 

I recommend using `vtk.js` to render in JavaScript first. This will help you avoid any problems that may arise from creating a Trame server. This can be achieved by specifying “client” in `set_jupyter_backend`. For more information, please refer to the following documentation.
https://docs.pyvista.org/version/stable/user-guide/jupyter/index.html#pyvista.set_jupyter_backend

```python
import pyvista

pyvista.start_xvfb()
pyvista.set_jupyter_backend("client")
```

There is also contents about backend of PyVista here.
https://tutorial.pyvista.org/tutorial/00_jupyter/index.html

Rendering using `vtk.js` does not fully support rendering in PyVista. However, it should be generally satisfactory.

A version of `vtk` with Wasm will be released soon, so if that comes to fruition, we will be able to provide full rendering in the client. For that reason, it is not very wise to struggle with building a `Trame` server at this point.

**Reply by stevejpurves** [reply]
Thank you for the help @tkoyama010 🙌

I've been trying this out and have a couple of questions:

* when setting `pyvista.set_jupyter_backend("client")` we would expect pyvista *not* to create an `iframe` pointed back to the user server url at all correct?

* is running `pyvista.start_xvfb()` essential for this approach to work? (if so it seems like i need to update my base image to `apt install` `libgl1-mesa-glx xvfb`) as with the following minimal example below, I don't get the `iframe` 404/issues but I do not get any output rendered either -- this is probably as myst has not got any pyvista/vtk frontend (rendermime) code loaded. So to enable client rendering we would need to go down that route...

**Reply by tkoyama010** [reply, marked as answer]
> * when setting `pyvista.set_jupyter_backend("client")` we would expect pyvista *not* to create an `iframe` pointed back to the user server url at all correct?

My bad. I thought that PyVista's client mode doesn't builds the server and iframe, but I was wrong :bow: I checked and it does call the build of the trame server even in `“client”` mode. Let me correct it to `“html”` mode, not `“client”` mode. Now no iframe url will be created.
```python
import pyvista

pyvista.start_xvfb()
pyvista.set_jupyter_backend("html")
```
Please see [set_jupyter_backend](https://docs.pyvista.org/version/stable/user-guide/jupyter/index.html#pyvista.set_jupyter_backend) document for more detail.
This option uses the `ipywidgets` to display html. If you do not want to use `ipywidgets` as well, please use the [export_html](https://docs.pyvista.org/version/stable/api/plotting/_autosummary/pyvista.Plotter.export_html.html#pyvista.Plotter.export_html) method of the Plotter class.

> is running `pyvista.start_xvfb()` essential for this approach to work? (if so it seems like i need to update my base image to `apt install` `libgl1-mesa-glx xvfb`) as with the following minimal example below, I don't get the `iframe` 404/issues but I do not get any output rendered either -- this is probably as myst has not got any pyvista/vtk frontend (rendermime) code loaded. So to enable client rendering we would need to go down that route...

Yes. That is also required in this case. (Edit: I have revised this response as my research has shown that it needs to be corrected.)

**Reply by tkoyama010** [reply]
If we use `vtk-osmesa` (see #4024), we may not need `pyvista.start_xvfb()`, but I don't check yet.
